Query Runner: eccenca Corporate Memory (SPARQL) - query RDF / Linked Data Knowledge Graphs with redash

seebi opened the pull request:
## What type of PR is this? (check all applicable)

- [x] New Query Runner (Data Source)

## Description

This adds a query runner for [eccenca Corporate Memory](https://documentation.eccenca.com/) based on the [SPARQL graph query language](https://www.w3.org/TR/sparql11-overview/).

The runner is in use internally at eccenca with redash 8 since 07/2020:
https://documentation.eccenca.com/blog/2020/07/redash-integration-ahead

The runner is based on our [python API package cmem-cmempy](https://pypi.org/project/cmem-cmempy/).

eccenca Corporate Memory is an Enterprise Knowledge Graph Platform which enables organisations to manage, build and consume RDF / Linked Data based Knowledge Graphs.

## Related Tickets & Documents

- https://www.w3.org/TR/sparql11-overview/
- https://documentation.eccenca.com/blog/2020/07/redash-integration-ahead
- https://pypi.org/project/cmem-cmempy/
- https://documentation.eccenca.com/

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

live action with redash main branch:
![2021-03-16 13 27 25](https://user-images.githubusercontent.com/510452/111309249-f70e7d80-865b-11eb-8e27-fffba1ff7702.gif)

screenshot based on redash 8:
<img width="1239" alt="2020-07-08-Redash-SPARQL-Query-Visualization" src="https://user-images.githubusercontent.com/510452/109996687-c18a9b80-7d0f-11eb-9b6b-71b41369189c.png">

seebi committed:
add Corporate Memory Runner based on cmempy 21.2.3

seebi committed:
fix code style

seebi committed:
apply some code nice ups

seebi commented:
@rauchy et al. : Any chance we can have a discussion about this PR?

arikfr commented on redash/query_runner/corporate_memory.py:
```diff
@@ -0,0 +1,257 @@
+            "properties": {
+                "CMEM_BASE_URI": {"type": "string", "title": "CMEM_BASE_URL"},
+                "OAUTH_GRANT_TYPE": {
+                    "type": "string",
+                    "title": "OAUTH_GRANT_TYPE (can be: password or client_credentials)",
+                    "default": "client_credentials",
```
We support dropdowns, so maybe better to use those to make sure people don't misspell?

seebi replied:
sure, exactly what is needed here. Will do

seebi replied:
@arikfr There quite some extensions to the schema description system compared to v8 (at lease I was not aware of it), so I not only introduced a dropdown but also a checkbox and the additional settings sections.

Hope this can be merged now.

btw: please also be aware of https://github.com/getredash/redash/pull/5415/commits/56349d4788d20d9026ca58519023c0f3d3fc4066

arikfr reviewed (commented):
Thank you for adding support for another data source! One comment on improving the configuration.

seebi committed:
use extendedEnum, boolean and extra_options for schema description

seebi committed:
use lower case sorting for data source types list

This correctly orders data source names which starts with lower
chars (such as eccenca Corporate Memory)

arikfr reviewed (approved):
Thanks!

arikfr commented:
I just noticed there is a conflict in requirements_all_ds.txt - can you please update?

seebi committed:
Merge remote-tracking branch 'upstream/master' into feature/cmemQueryRunner

seebi commented:
> I just noticed there is a conflict in requirements_all_ds.txt - can you please update?

done

seebi commented:
@arikfr any news on this? :-)

seebi committed:
add missing dblogo

arikfr commented:
👍 merged. Thanks!
